tools/codegen/gen.py

```diff
@@ -276,10 +276,21 @@ def func(f: NativeFunction) -> Optional[str]:
                     cuda_guard = """\
     // DeviceGuard omitted
 """
+            mem_overlap_check = ""
+            if (dispatch == "CPU" or dispatch == "CUDA") and ('backward' not in impl_name):
+                if f.func.name.name.inplace:
+                    mem_overlap_check = """\
+    at::assert_no_internal_overlap(self);
+"""
+                if f.func.is_out_fn():
+                    for out_argument in f.func.out_arguments:
+                        mem_overlap_check += f"""\
+    at::assert_no_internal_overlap({out_argument.name});
```

Comment:
This seems good. I don't know if there are any out= functions that always write the same value to the Tensor (if there are, then this would break BC)